src/Check/BestPracticesFolderStructure.php

```diff
@@ -49,59 +69,120 @@ public function getResultFail() {}
+  public function getResultInfo() {
+    return $this->infoMessage;
+  }
 
   /**
    * {@inheritdoc}.
    */
   public function getResultPass() {
-    return $this->t('modules/contrib and modules/custom directories exist.');
+    return $this->passMessage;
   }
 
   /**
    * {@inheritdoc}.
    */
   public function getResultWarn() {
-    if (!$this->registry->contrib && !$this->registry->custom) {
-      return $this->t('Neither modules/contrib nor modules/custom directories are present!');
-    }
-    if (!$this->registry->contrib) {
-      return $this->t('modules/contrib directory is not present!');
-    }
-    if (!$this->registry->custom) {
-      return $this->t('modules/custom directory is not present!');
-    }
+    return $this->warningMessage;
   }
 
   /**
    * {@inheritdoc}.
    */
   public function getAction() {
-    $message = '';
-    if ($this->score == SiteAuditCheckBase::AUDIT_CHECK_SCORE_WARN) {
-      if (!$this->registry->contrib && !$this->registry->custom) {
-        $message .= $this->t('Put all the contrib modules inside the ./modules/contrib directory and custom modules inside the ./modules/custom directory.');
-      }
-      elseif (!$this->registry->contrib) {
-        $message .= $this->t('Put all the contrib modules inside the ./modules/contrib directory.');
-      }
-      elseif (!$this->registry->custom) {
-        $message .= $this->t('Put all the custom modules inside the ./modules/custom directory.');
-      }
-      return $message . ' ' . $this->t('Moving modules may cause errors, so refer to https://www.drupal.org/node/183681 for information on how to best proceed.');
+    if (!$this->actionMessage) {
+      return null;
     }
+
+    return $this->actionMessage . ' ' . $this->t('Moving modules may cause errors, so refer to https://www.drupal.org/node/183681 for information on how to best proceed.');
   }
 
   /**
    * {@inheritdoc}.
    */
   public function calculateScore() {
-    $this->registry->contrib = is_dir(DRUPAL_ROOT . '/modules/contrib');
-    $this->registry->custom = is_dir(DRUPAL_ROOT . '/modules/custom');
-    if (!$this->registry->contrib || !$this->registry->custom) {
+    $subDirs = glob(
+        sprintf('%s/modules/*', DRUPAL_ROOT),
+        GLOB_ONLYDIR
+    );
+    if (!$subDirs) {
+      // No subdirectories inside "modules/" directory found.
+      $this->infoMessage = $this->t('Contrib and custom modules not found.');
+
+      return SiteAuditCheckBase::AUDIT_CHECK_SCORE_INFO;
+    }
+
+    $validContribModulesDir = $this->getValidDirName($subDirs, array('contrib', 'composer'));
+    $validCustomModulesDir = $this->getValidDirName($subDirs, array('custom'));
+    if (1 === count($subDirs)) {
+      // Only one subdirectory inside "modules/" directory found.
+      if ($validContribModulesDir || $validCustomModulesDir) {
+        $this->passMessage = $this->t(
+          'modules/@subdir directory exist.',
+          array('@subdir' => basename($subDirs[0]))
```

Comment:
In this instance, let's also advise the user to upgrade to drupal-composer-managed if the valid contrib modules dir is named `composer`. This is something of a Pantheon-specific message in a non-Panthoen-specific bit of code; however, the odds that anyone would use a subdir named `composer`, but **not** get that name from pantheon-upstreams/drupal-project is very low, so I think this should be okay.